File: simulate.py
import random
import sys

CUT = 0.5

def simulate(n, limit, maxChildren):
	species = []
	firstGeneration = [' ' + chr(65 + i) for i in range(0, limit)]
	species.append(firstGeneration)
	for i in range(0, n-1):
		species.append(evolute(species[i], limit, maxChildren))
	return species

def evolute(generation, limit, maxChildren):
	children = {}
	allChild = 0

	outerLoop = True
	while outerLoop:
		list = [ { 'index': i, 'value': random.random() } for i in range(0, len(generation))]
		toBirth = filter(lambda x: x['value'] > CUT, list)
		toBirth.sort(lambda x, y: int(x['value'] - y['value']))
		for tb in toBirth:
			idx = tb['index']
			if not children.has_key(idx):
				children[idx] = 0
			if children[idx] == maxChildren:
				continue
			children[idx] += 1
			allChild += 1
			if allChild == limit:
				outerLoop = False
				break

	nextGeneration = []
	j = 0
	for i in range(0, limit):
		individual = generation[i]
		if not children.has_key(i):
			continue
		parent = individual[-1]
		j0 = j
		while j - j0 < children[i]:
			nextGeneration.append(parent + chr(65 + j))
			j += 1

	return nextGeneration

def main(argv):
	n = int(argv[0] if len(argv) > 1 else 100)
	limit = int(argv[1] if len(argv) > 2 else 20)
	maxChildren = int(argv[2] if len(argv) > 3 else 3)
	species = simulate(n, limit, maxChildren)
	text = '\r\n'.join(map(lambda x: ' '.join(x), species))
	print(text)

if __name__ == '__main__':
	main(sys.argv[1:])

File: simulate.js

if (require.main == module) {
	var argv = process.argv.splice(2);
	main(argv);
}

const CUT = 0.5;

function main(argv) {
	var n = parseInt(argv[0] || 100);
	var limit = parseInt(argv[1] || 20);
	var maxChildren = parseInt(argv[2] || 3);
	var species = simulate(n, limit, maxChildren);
	var text = species.map(x => x.join(' ')).join('\r\n');
	console.log(text);
}

function simulate(n, limit, maxChildren) {
	var species = [];
	species.push(initCreate(limit));
	for (var i=0; i<n-1; i++)
		species.push(evolute(species[i], limit, maxChildren));
	return species;
}

function initCreate(limit) {
	var generation = [];
	for (var i=0; i<limit; i++)
		generation[i] = ' ' + String.fromCharCode(65 + i);
	return generation;
}

function evolute(generation, limit, maxChildren) {
	var children = {};
	var allChild = 0;

outter:
	while (true) {
		var list = [];
		for (var i=0; i<generation.length; i++) {
			list.push({
				index: i,
				value: Math.random()
			});
		}
		var toBirth = list.filter(x => x.value > CUT).sort((x, y) => y.value - x.value);
		for (var i=0; i<toBirth.length; i++) {
			var tb = toBirth[i];
			if (typeof children[tb.index] === 'undefined')
				children[tb.index] = 0;
			if (children[tb.index] == maxChildren) continue;
			children[tb.index]++;
			allChild++;
			if (allChild == limit) break outter;
		}
	}

	var nextGeneration = [];
	var j = 0;
	for (var i=0; i<generation.length; i++) {
		if (typeof children[i] === 'undefined') continue;
		var parent = generation[i].substr(generation[i].length - 1);
		var j0 = j;
		while (j - j0 < children[i])
			nextGeneration[j] = parent + String.fromCharCode(65 + j++);
	}

	return nextGeneration;
}
